presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -60,220 +64,149 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
-                return Optional.empty();
+            if (canReplaceWithRowNumber(node)) {
+                return new RowNumberNode(idAllocator.getNextId(),
+                        rewrittenSource,
+                        node.getPartitionBy(),
+                        getOnlyElement(node.getWindowFunctions().keySet()),
+                        Optional.empty(),
+                        Optional.empty());
             }
-            if (filter.getLimit().isPresent()) {
-                return filter.getLimit();
-            }
-            if (filterContainsWindowFunctions(node, filter.getFilterExpression().get()) &&
-                    filter.getFilterExpression().get() instanceof ComparisonExpression) {
-                Symbol rowNumberSymbol = Iterables.getOnlyElement(node.getWindowFunctions().entrySet()).getKey();
-                return WindowLimitExtractor.extract(filter.getFilterExpression().get(), rowNumberSymbol);
-            }
-            return Optional.empty();
-        }
-
-        private static boolean canOptimizeWindowFunction(WindowNode node)
-        {
-            if (node.getWindowFunctions().size() != 1) {
-                return false;
-            }
-            Symbol rowNumberSymbol = getOnlyElement(node.getWindowFunctions().entrySet()).getKey();
-            return isRowNumberSignature(node.getSignatures().get(rowNumberSymbol));
-        }
-
-        private static boolean filterContainsWindowFunctions(WindowNode node, Expression filterPredicate)
-        {
-            Set<Symbol> windowFunctionSymbols = node.getWindowFunctions().keySet();
-            Sets.SetView<Symbol> commonSymbols = Sets.intersection(DependencyExtractor.extractUnique(filterPredicate), windowFunctionSymbols);
-            return !commonSymbols.isEmpty();
+            return context.replaceChildren(node, ImmutableList.of(rewrittenSource));
         }
 
         @Override
-        public PlanNode visitLimit(LimitNode node, RewriteContext<Constraint> context)
+        public PlanNode visitLimit(LimitNode node, RewriteContext<Void> context)
         {
             // Operators can handle MAX_VALUE rows per page, so do not optimize if count is greater than this value
             if (node.getCount() >= Integer.MAX_VALUE) {
                 return context.defaultRewrite(node);
             }
-            Constraint constraint = new Constraint(Optional.of((int) node.getCount()), Optional.empty());
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), constraint);
 
-            if (rewrittenSource != node.getSource()) {
-                return new LimitNode(idAllocator.getNextId(), rewrittenSource, node.getCount());
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            if (rewrittenSource instanceof RowNumberNode) {
+                RowNumberNode rowNumberNode = (RowNumberNode) rewrittenSource;
+                if (rowNumberNode.getPartitionBy().isEmpty()) {
```

Comment:
Not something you need to change right now, but I still find it a little weird that we only want to optimize the case where the partitionBy is empty. From what I can tell, you should be able to push it in regardless. And in the case where partition By is not empty, you can just check the limit again.

On the other hand, the RowNumber operator is already pretty efficient in that it is streaming through the pages. In which case, evaluating the limit after a row number without pushing it in might be simpler and still just as fast (you will only over-process by at most 1 page of data).